<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>iMIRAGE Vignette • iMIRAGE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="iMIRAGE Vignette">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">iMIRAGE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/imirage.html">iMIRAGE Vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>iMIRAGE Vignette</h1>
                        <h4 class="author">Aritro Nath</h4>
            
            <h4 class="date">2019-08-01</h4>
      
      
      <div class="hidden name"><code>imirage.Rmd</code></div>

    </div>

    
    
<hr>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><strong>iMIRAGE</strong> stands for <em>imputed microRNA (miRNA) activity from gene expression</em>. As the name suggests, the package imputes expression of miRNAs by constructing prediction models that only depend on the expression levels of protein-coding genes. In essence, iMIRAGE package can impute the miRNA profiles of samples where protein-coding expression data is available (for example, from microarray or RNA-seq), but do not contain reliable miRNA expression. By utilizing a <em>training</em> data set containing both protein-coding and miRNA expression profiles, iMIRAGE constructs prediction models using machine-learning to impute miRNA profiles of the <em>independent</em> <em>(test)</em> data set of interest.</p>
<p>The iMIRAGE package also provides tools to create an integrated workflow, to harmonize, clean-up, normalize and standardize the expression data sets. In addition, the package provides an option of using miRNA-target gene pair information to construct the prediction models</p>
<div id="download-and-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#download-and-installation" class="anchor"></a>Download and installation</h2>
<p>The iMIRAGE package for R can be downloaded from the <a href="https://github.com/aritronath/iMIRAGE">GitHub repository</a>:</p>
<pre><code><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">#Download package from GitHub using devtools
devtools::install_github("aritronath/iMIRAGE")</a></code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Load the package in R</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(iMIRAGE)
<span class="co">#&gt; Loading required package: randomForest</span>
<span class="co">#&gt; randomForest 4.6-14</span>
<span class="co">#&gt; Type rfNews() to see new features/changes/bug fixes.</span>
<span class="co">#&gt; Loading required package: FNN</span>
<span class="co">#&gt; Loading required package: e1071</span></code></pre></div>
</div>
<div id="whats-included-in-this-package" class="section level2">
<h2 class="hasAnchor">
<a href="#whats-included-in-this-package" class="anchor"></a>Whats included in this package?</h2>
<p>The iMIRAGE package includes necessary functions for pre-processing and harmonization of expression data sets, miRNA expression imputation and cross-validation analyses. In addition, TargetScan miRNA-target gene pairs are provided for constructing imputation models. The package also includes the example data sets that are used in this vignette.</p>
</div>
<div id="report-issues" class="section level2">
<h2 class="hasAnchor">
<a href="#report-issues" class="anchor"></a>Report issues</h2>
<p><a href="https://github.com/aritronath/iMIRAGE/issues">Report bugs and issues here via GitHub</a></p>
<hr>
</div>
</div>
<div id="quick-start-guide" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start-guide" class="anchor"></a>Quick Start Guide</h1>
<p>In this example, we will use the two <strong>independent</strong> miRNA data sets that were derived from the Cancer Genome Atlas (TCGA) Breast Cancer (BRCA) project. These data sets are automatically available when the iMIRAGE package library is loaded.</p>
<div id="datasets" class="section level3">
<h3 class="hasAnchor">
<a href="#datasets" class="anchor"></a>Datasets:</h3>
<ol style="list-style-type: decimal">
<li>Training: <em>GA.pcg</em>, <em>GA.mir</em>
<ul>
<li>The prefix <em>GA</em> refers to TCGA breast invasive carcinoma (BRCA) samples for which miRNA mature strand expression were obtained using the Illumina Genome Analyzer system.</li>
</ul>
</li>
<li>Test: <em>HS.pcg</em>, <em>HS.mir</em>
<ul>
<li><p>The prefix <em>HS</em> refers to TCGA BRCA samples for which miRNA mature strand expression were obtained using the Illumina Hiseq system.</p></li>
<li><p>The samples in the two data sets are mutually exclusive. The small subset included in these example data sets were randomly selected</p></li>
<li><p>Note: the complete TCGA BRCA HiSeq and GA data sets can be downloaded from the <a href="https://xenabrowser.net/datapages/?cohort=TCGA%20Breast%20Cancer%20(BRCA)&amp;removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443">XENA portal</a></p></li>
</ul>
</li>
</ol>
</div>
<div id="step-1-match-protein-coding-genes-predictive-features-in-the-training-and-test-datasets-with-match-gex" class="section level2">
<h2 class="hasAnchor">
<a href="#step-1-match-protein-coding-genes-predictive-features-in-the-training-and-test-datasets-with-match-gex" class="anchor"></a>Step 1: Match protein-coding genes (predictive features) in the training and test datasets with <code>match.gex</code>
</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># return a pair of matrices with matched columns (protein-coding genes that will be used as training features)</span>

temp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match.gex.html">match.gex</a></span>(GA.pcg, HS.pcg)
GA.pcg &lt;-<span class="st"> </span>temp[[<span class="dv">1</span>]]
HS.pcg &lt;-<span class="st"> </span>temp[[<span class="dv">2</span>]]</code></pre></div>
</div>
<div id="step-2a-perform-cross-validation-analysis-with-imirage-cv-to-determine-the-expected-accuracy-of-imputing-a-mirna-of-interest-using-the-training-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#step-2a-perform-cross-validation-analysis-with-imirage-cv-to-determine-the-expected-accuracy-of-imputing-a-mirna-of-interest-using-the-training-datasets" class="anchor"></a>Step 2A: Perform cross-validation analysis with <code>imirage.cv</code> to determine the expected accuracy of imputing a miRNA of interest using the training datasets</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Use **imirage.cv** with the default parameters of using K-nearest neighbors as the prediction algorithm and using 50 protein-coding genes as training features</span>

CV.miRNA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/imirage.cv.html">imirage.cv</a></span>(<span class="dt">train_pcg=</span>GA.pcg, <span class="dt">train_mir=</span>GA.mir, <span class="dt">gene_index=</span><span class="st">"hsa-let-7c"</span>, <span class="dt">method=</span><span class="st">"KNN"</span>, <span class="dt">target=</span><span class="st">"none"</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; Running 10-folds cross-validation...</span>
<span class="co">#&gt; Iteration 1</span>
<span class="co">#&gt; Iteration 2</span>
<span class="co">#&gt; Iteration 3</span>
<span class="co">#&gt; Iteration 4</span>
<span class="co">#&gt; Iteration 5</span>
<span class="co">#&gt; Iteration 6</span>
<span class="co">#&gt; Iteration 7</span>
<span class="co">#&gt; Iteration 8</span>
<span class="co">#&gt; Iteration 9</span>
<span class="co">#&gt; Iteration 10</span>
<span class="co">#&gt; Cross-validation complete</span>

<span class="co">#Print the accuracy metrics from each fold of the cross-validation analysis </span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(CV.miRNA)
<span class="co">#&gt;             PCC      P-Value      RMSE</span>
<span class="co">#&gt;  [1,] 0.6421569 6.652233e-03 1607.9770</span>
<span class="co">#&gt;  [2,] 0.7203907 3.688712e-05 1369.3719</span>
<span class="co">#&gt;  [3,] 0.6840176 2.602183e-05 1467.2613</span>
<span class="co">#&gt;  [4,] 0.6628352 1.744676e-04 1824.5160</span>
<span class="co">#&gt;  [5,] 0.7235963 4.309439e-06 1898.4681</span>
<span class="co">#&gt;  [6,] 0.7512333 5.235529e-08 1811.2824</span>
<span class="co">#&gt;  [7,] 0.6729700 7.005981e-05 2697.4273</span>
<span class="co">#&gt;  [8,] 0.7908907 9.076999e-08 2199.1467</span>
<span class="co">#&gt;  [9,] 0.6849761 2.467588e-07  827.1981</span>
<span class="co">#&gt; [10,] 0.7393048 2.451796e-06 1376.7202</span>

<span class="co">#note: this example performs cross-validation analysis for 1 unique miRNA, hsa-let-7c. The name must match with one of the names in the train_mir object's column names</span></code></pre></div>
</div>
<div id="step-2b-perform-cross-validation-analysis-with-imirage-cv-loop-to-determine-the-expected-imputation-accuracy-of-entire-mirna-dataset-with" class="section level2">
<h2 class="hasAnchor">
<a href="#step-2b-perform-cross-validation-analysis-with-imirage-cv-loop-to-determine-the-expected-imputation-accuracy-of-entire-mirna-dataset-with" class="anchor"></a>Step 2B: Perform cross-validation analysis with <code>imirage.cv.loop</code> to determine the expected imputation accuracy of entire miRNA dataset with</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Perform cross-validation analysis over the entire training dataset </span>
CV.full &lt;-<span class="st"> </span><span class="kw"><a href="../reference/imirage.cv.loop.html">imirage.cv.loop</a></span>(<span class="dt">train_pcg =</span> GA.pcg, <span class="dt">train_mir =</span> GA.mir, <span class="dt">method =</span> <span class="st">"KNN"</span>, <span class="dt">target=</span><span class="st">"none"</span>)

<span class="co">#Plot some performance metrics from cross-validation analysis</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(CV.full[,<span class="dv">1</span>], <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(CV.full[,<span class="dv">2</span>]), <span class="dt">xlab=</span><span class="st">"Spearman R"</span>, <span class="dt">ylab=</span><span class="st">"P-value (-log10)"</span>, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">main=</span><span class="st">"Cross-validation accuracy"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/abline">abline</a></span>(<span class="dt">h=</span><span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(<span class="fl">0.05</span>), <span class="dt">lty=</span><span class="dv">3</span>)</code></pre></div>
<p><img src="imirage_files/figure-html/unnamed-chunk-4-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Find out which miRNAs are imputed with good accuracy </span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(GA.mir)[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(CV.full[,<span class="dv">1</span>] <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>)] <span class="co">#arbritarily, Spearman correlation &gt; 0.5 </span>
<span class="co">#&gt;  [1] "hsa-let-7c"   "hsa-mir-100"  "hsa-mir-10b"  "hsa-mir-1301"</span>
<span class="co">#&gt;  [5] "hsa-mir-1307" "hsa-mir-130b" "hsa-mir-141"  "hsa-mir-150" </span>
<span class="co">#&gt;  [9] "hsa-mir-17"   "hsa-mir-18a"  "hsa-mir-190b" "hsa-mir-20a" </span>
<span class="co">#&gt; [13] "hsa-mir-210"  "hsa-mir-301a" "hsa-mir-3677" "hsa-mir-4772"</span>
<span class="co">#&gt; [17] "hsa-mir-99a"</span>

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(GA.mir)[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(CV.full[,<span class="dv">2</span>] <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)] <span class="co">#P-value &lt; 0.05</span>
<span class="co">#&gt;  [1] "hsa-let-7c"   "hsa-mir-100"  "hsa-mir-10b"  "hsa-mir-1301"</span>
<span class="co">#&gt;  [5] "hsa-mir-1307" "hsa-mir-130b" "hsa-mir-141"  "hsa-mir-150" </span>
<span class="co">#&gt;  [9] "hsa-mir-155"  "hsa-mir-17"   "hsa-mir-18a"  "hsa-mir-190b"</span>
<span class="co">#&gt; [13] "hsa-mir-20a"  "hsa-mir-210"  "hsa-mir-301a" "hsa-mir-3677"</span>
<span class="co">#&gt; [17] "hsa-mir-4772" "hsa-mir-99a"</span></code></pre></div>
</div>
<div id="step-3a-impute-expression-of-a-single-mirna-of-interest-using-imirage-in-the-test-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#step-3a-impute-expression-of-a-single-mirna-of-interest-using-imirage-in-the-test-dataset" class="anchor"></a>Step 3A: Impute expression of a single miRNA of interest using <code>imirage</code> in the test dataset</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Use **imirage** with default parameters</span>
Pred.miRNA &lt;-<span class="st"> </span><span class="kw"><a href="../reference/imirage.html">imirage</a></span>(<span class="dt">train_pcg=</span>GA.pcg, <span class="dt">train_mir=</span>GA.mir, <span class="dt">my_pcg=</span>HS.pcg , <span class="dt">gene_index=</span><span class="st">"hsa-let-7c"</span>, <span class="dt">target=</span><span class="st">"none"</span>)

<span class="co">#Display the predicted miRNA expression values</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(Pred.miRNA)
<span class="co">#&gt;  [1] 2019.3563 1498.5680 1771.0472 2525.3074 3406.9625 1813.6320 3009.6051</span>
<span class="co">#&gt;  [8] 2198.6933 3072.8239 1307.7639 2962.8586 1277.5310 2258.6073 2110.7417</span>
<span class="co">#&gt; [15] 5194.3184 2963.7463 1765.5263 3669.1882 1952.3878 3368.4370 1316.7190</span>
<span class="co">#&gt; [22] 2017.3545 1671.8867 5657.6180 1037.8135 3214.5502 5665.0500 1815.9127</span>
<span class="co">#&gt; [29] 2612.9871 2567.6086 2341.5214 2787.6283 2312.1913 1230.4309 2020.4602</span>
<span class="co">#&gt; [36]  969.7257 1145.1208  958.9258 1043.2435 1181.7525 1092.2694 2585.8381</span>
<span class="co">#&gt; [43] 4762.9124 1914.3931 1659.7709 2286.6204 1590.2248  987.3592 1146.6056</span>
<span class="co">#&gt; [50] 1086.8198</span>

<span class="co">#Compare the predicted miRNA expression values with measured expression </span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(Pred.miRNA, HS.mir[,<span class="st">"hsa-let-7c"</span>], <span class="dt">xlab=</span><span class="st">"Predicted"</span>, <span class="dt">ylab=</span><span class="st">"Measured"</span>, <span class="dt">main=</span><span class="st">"hsa-let-7c imputation performance"</span>, <span class="dt">pch=</span><span class="dv">16</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/abline">abline</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/lm">lm</a></span>(HS.mir[,<span class="st">"hsa-let-7c"</span>] <span class="op">~</span><span class="st"> </span>Pred.miRNA), <span class="dt">lty=</span><span class="dv">3</span>)</code></pre></div>
<p><img src="imirage_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
</div>
<div id="step-3b-use-imirage-to-impute-expression-of-all-mirnas-available-in-the-training-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#step-3b-use-imirage-to-impute-expression-of-all-mirnas-available-in-the-training-dataset" class="anchor"></a>Step 3B: Use <code>imirage</code> to impute expression of all miRNAs available in the training dataset</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Create an empty matrix to store imputed expression </span>
Pred.full &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data=</span><span class="ot">NA</span>, <span class="dt">ncol=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(GA.mir), <span class="dt">nrow=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(HS.pcg))

<span class="co">#Execute a loop to impute each miRNA using the test protein coding dataset</span>
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(GA.mir)) {
  Pred.full[,i] &lt;-<span class="st"> </span><span class="kw"><a href="../reference/imirage.html">imirage</a></span>(<span class="dt">train_pcg =</span> GA.pcg, <span class="dt">train_mir =</span> GA.mir, <span class="dt">my_pcg =</span> HS.pcg, <span class="dt">gene_index =</span> i, <span class="dt">method=</span><span class="st">"KNN"</span>, <span class="dt">target=</span><span class="st">"none"</span>)
}</code></pre></div>
<p>Find out how well can we impute miRNA expression in the <strong>independent</strong> data set</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Obtain correlation coefficients between imputed and measured miRNA expression </span>
Pred.Cors &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/array">array</a></span>(<span class="dt">dim=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(GA.mir))
<span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(GA.mir)) {
  Pred.Cors[i] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(HS.mir[,i], Pred.full[,i], <span class="dt">method=</span><span class="st">"spearman"</span>)
}

<span class="co">#Plot imputation correlations in comparison to cross-validation results</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(Pred.Cors, CV.full[,<span class="dv">1</span>], <span class="dt">xlab=</span><span class="st">"Imputation accuracy"</span>, <span class="dt">ylab=</span><span class="st">"Cross-validation accuracy"</span>, <span class="dt">main=</span><span class="st">"Imputation performance"</span>, <span class="dt">pch=</span><span class="dv">16</span>)</code></pre></div>
<p><img src="imirage_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<hr>
</div>
</div>
<div id="detailed-guide" class="section level1">
<h1 class="hasAnchor">
<a href="#detailed-guide" class="anchor"></a>Detailed guide</h1>
<div id="background" class="section level2">
<h2 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h2>
<p>The principle behind iMIRAGE can be explained in a simplified form by stating the problem in the form of a linear model. Let us assume the user wants to obtain miRNA expression using their protein-coding data set which we will call <em>test</em>. To impute the miRNA expression using the <em>test</em> protein-coding data set, we first determine the mathematical relation between the expression of a set of miRNAs <em>Y</em> with the set of protein-coding gene expression <em>X</em> of the <em>training</em> data set, where <em>p</em> indicates the number of protein-coding genes used in the model. This number can range anywhere from 1 to <em>P</em>, where, <em>P = total number of protein-coding genes</em> available in both the training and test data sets.</p>
<p><span class="math inline">\(Y^{Training} = \sum_{i=1}^{p} \beta X_{i}^{Training} + \epsilon\)</span></p>
<p>The objective of iMIRAGE is to determine <span class="math inline">\(\beta\)</span> so that it can be harnessed to predict the expression of miRNA <em>Y</em> by applying the model to the <em>test</em> protein-coding data set:</p>
<p><span class="math inline">\(Y^{Predicted} = \sum_{i=1}^{p} \beta X_{i}^{Test}\)</span></p>
<p>(Note that the machine-learning algorithms that are used iMIRAGE are not necessarily based on the assumption of linearity. This is only for simplified illustration purposes)</p>
<p>iMIRAGE assumes that not all protein-coding genes are meaningful for constructing imputation models. Therefore, it suggests using a subset of informative protein-coding genes for each miRNA. These informative features are automatically determined by calculating the correlation between the expression of all features with miRNA expression in the training data sets. Then, only a user-specified subset of the top correlated features are retained for constructing the models (by default 50).</p>
</div>
<div id="required-data" class="section level2">
<h2 class="hasAnchor">
<a href="#required-data" class="anchor"></a>Required data</h2>
<ol style="list-style-type: decimal">
<li>Training data: protein-coding and miRNA expression data. In the package documentation, these data sets are referred by their alias <em>train_pcg</em> and <em>train_mir</em> respectively
<ul>
<li>Both data sets must be from the same samples</li>
<li>High-quality miRNA expression data - from small RNA-seq or miRNA arrays preferred</li>
<li>miRNA expression from regular RNA-seq libraries are not reliable</li>
<li>The minimum recommended sample size for training datasets is 50</li>
</ul>
</li>
<li><p>Protein-coding gene expression data from the independent samples of interest. In the package documentation, this data set is referred by its alias <em>my_pcg</em></p></li>
<li><p>Ideally, the training dataset should be of the same tissue-type as the independent test dataset. However, large datasets spanning multiple tissues can also be used. For example, <a href="https://portal.gdc.cancer.gov/">the pan-cancer TCGA</a> cohort covering more than 30 tissue-types can be used as training data to predict miRNA profiles of various individual cancers.</p></li>
<li>Organizing your data before use:
<ul>
<li>Your data should be arranged as a <em>n x p</em> matrix, where <em>n</em> denotes samples in <em>rows</em> and <em>p</em> denotes genes/miRNA in <em>columns</em>
</li>
<li>Row names should be names of the samples</li>
<li>Column names should be names of the gene/miRNA</li>
<li>Gene IDs should be of the same type in the training and test data sets</li>
<li>If you plan to use target gene pairs for constructing imputation models, the miRNA nomenclature should match between the training miRNA data set and the miRNA-target gene pairs. Similarly, the protein-coding gene IDs in the training and test data sets should match with the type of ID used by the miRNA-target gene pair data set</li>
<li>To convert gene IDs, use the R/Bioconductor package <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a>
</li>
</ul>
<pre><code>#Short example on converting RefSeq mRNA IDs to ENSEMBL IDs using biomaRt
library(biomaRt)
ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")
old.ids &lt;- colnames(my_pcg) #RefSeq IDs
gene.annot &lt;- getBM(attributes=c('refseq_mrna','ensembl_gene_id'), 
             filters = 'refseq_mrna', 
             values = as.character(old.ids), 
             mart = ensembl) 
#returns a dataframe with ENSEMBL IDs corresponding to supplied RefSeq mRNA IDs</code></pre>
<p><a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/biomaRt/inst/doc/biomaRt.html">See the biomaRt documentation for complete details and instructions</a></p>
</li>
</ol>
</div>
<div id="workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow" class="anchor"></a>Workflow</h2>
<p>We will use the unprocessed TCGA BRCA data sets to illustrate a typical workflow. The data set used in the subsequent examples can be downloaded from the Open Science Framework repository by following the following URL: <a href="https://osf.io/s5rbn/download" class="uri">https://osf.io/s5rbn/download</a></p>
<p>After downloading the TCGA_BRCA_Datasets.RData file in the current working directory, load the contents using</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/load">load</a></span>(<span class="st">"TCGA_BRCA_Datasets.RData"</span>)</code></pre></div>
<p>The following four matrices should appear after the object is loaded: ga.gex, ga.mirna, hiseq.gex, hiseq.mirna</p>
<div id="filtering-training-datasets-using-filter-expto-remove-gene-or-mirnas-that-are-not-expressed-in-most-samples-using-optional" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-training-datasets-using-filter-expto-remove-gene-or-mirnas-that-are-not-expressed-in-most-samples-using-optional" class="anchor"></a>1. Filtering training datasets using <code>filter.exp</code>to remove gene or miRNAs that are not expressed in most samples using (optional)</h3>
<p>We first remove gene or miRNAs that are not expressed in most samples, as these genes are likely not measured due to technical limitations and may not be reliable expression estimates. Ultimately, it is up to the user to decide whether they would like to include the sparse genes or miRNA in their analyses.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Here, we filter the miRNA datasets to retain all miRNAs that are expressed above a level of 0 in atleast 75% of the samples</span>
ga.mirna &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filter.exp.html">filter.exp</a></span>(ga.mirna, <span class="dt">cutoff=</span><span class="dv">75</span>, <span class="dt">threshold =</span> <span class="dv">0</span>)
hiseq.mirna &lt;-<span class="st"> </span><span class="kw"><a href="../reference/filter.exp.html">filter.exp</a></span>(hiseq.mirna, <span class="dt">cutoff=</span><span class="dv">75</span>, <span class="dt">threshold =</span> <span class="dv">0</span>)</code></pre></div>
</div>
<div id="use-match-gex-to-match-training-and-test-protein-coding-expression-matrices-required" class="section level3">
<h3 class="hasAnchor">
<a href="#use-match-gex-to-match-training-and-test-protein-coding-expression-matrices-required" class="anchor"></a>2. Use <code>match.gex</code> to match training and test protein-coding expression matrices (required)</h3>
<p>Here, we match the two protein-coding expression matrices for subsequent processing</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#We also will keep the unprocessed hiseq.gex.1 and ga.gex.1 datasets for subsquent comparisons</span>
temp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match.gex.html">match.gex</a></span>(hiseq.gex, ga.gex)
hiseq.gex.<span class="dv">1</span> &lt;-<span class="st"> </span>temp[[<span class="dv">1</span>]]
ga.gex.<span class="dv">1</span> &lt;-<span class="st"> </span>temp[[<span class="dv">2</span>]]

<span class="co">#For subsquent prediction analyses, we will also select miRNAs are common to both datasets</span>
temp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match.gex.html">match.gex</a></span>(ga.mirna, hiseq.mirna)
ga.mirna.<span class="dv">1</span> &lt;-<span class="st"> </span>temp[[<span class="dv">1</span>]]
hiseq.mirna.<span class="dv">1</span> &lt;-<span class="st"> </span>temp[[<span class="dv">2</span>]]</code></pre></div>
</div>
<div id="pre-process-training-and-test-datasets-optional-recommended" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-process-training-and-test-datasets-optional-recommended" class="anchor"></a>3. <code>pre.process</code> training and test datasets (optional, recommended)</h3>
<p>In this step, we process the training and test data sets using <code>pre.process</code> to make sure they are transformed, normalized and standardized before proceeding with further analysis. These steps tend to have a significant impact on subsequent analyses.</p>
<p>Generally, the user should determine whether their data has been transformed (<span class="math inline">\(\log2\)</span> or <span class="math inline">\(\log2(x+1)\)</span>) or normalized from the original source. In most cases, microarray data sets in public domain, such as <a href="https://www.ncbi.nlm.nih.gov/geo/">NCBI GEO</a> or <a href="https://www.ebi.ac.uk/arrayexpress/">ArrayExpress</a>, are already transformed and normalized. However, several RNA-seq data sets may be available prior to transformation and normalization. In these instances, it is advisable to perform log transformation and upper-quantile normalization prior to use in the workflow.</p>
<p>In addition, the variance filter removes genes with zero variance that will offer no predictive power. Finally, the data is scaled to a mean = 0 and standard deviation = 1. These are bare-minimum standardization techniques which can be skipped if the user has already pre-processed their data using other approaches.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ga.gex.<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pre.process.html">pre.process</a></span>(ga.gex.<span class="dv">1</span>, <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">var.filter =</span> <span class="ot">TRUE</span>, <span class="dt">UQ =</span> <span class="ot">TRUE</span>, <span class="dt">std =</span> <span class="ot">TRUE</span>)
hiseq.gex.<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pre.process.html">pre.process</a></span>(hiseq.gex.<span class="dv">1</span>, <span class="dt">log =</span> <span class="ot">TRUE</span>, <span class="dt">var.filter =</span> <span class="ot">TRUE</span>, <span class="dt">UQ =</span> <span class="ot">TRUE</span>, <span class="dt">std =</span> <span class="ot">TRUE</span>)

<span class="co">#This is how the data looks before and after pre.processing (showing the first 10 samples from each dataset)</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/par">par</a></span>(<span class="dt">mfrow=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">cex=</span><span class="fl">0.75</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/boxplot">boxplot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(ga.gex.<span class="dv">1</span>[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,]), <span class="dt">main=</span><span class="st">"GA - raw expression"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/boxplot">boxplot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(ga.gex.<span class="dv">2</span>[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,]), <span class="dt">main=</span><span class="st">"GA - processed expression"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/boxplot">boxplot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(hiseq.gex.<span class="dv">1</span>[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,]), <span class="dt">main=</span><span class="st">"Hiseq - raw expression"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/boxplot">boxplot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(hiseq.gex.<span class="dv">2</span>[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,]), <span class="dt">main=</span><span class="st">"Hiseq - processed expression"</span>)</code></pre></div>
<p><img src="imirage_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
</div>
<div id="performing-cross-validation-to-obtain-accuracy-metrics-and-imputing-in-test-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#performing-cross-validation-to-obtain-accuracy-metrics-and-imputing-in-test-dataset" class="anchor"></a>4. Performing cross-validation to obtain accuracy metrics and imputing in test dataset</h3>
<p>Here, we will first find out which miRNAs can be imputed with good accuracy using the GA data set as our training data set. The <code>imirage.cv</code> function performs a 10-fold cross-validation analysis for a single miRNA specified by the <em>gene_index</em> argument, as shown in the quick start guide above. In addition, the <code>imirage.cv.loop</code> wrapper can be used to perform the cross-validation analysis on the entire training miRNA data set.</p>
<p>The cross-validation analysis can be performed using one of the three methods: K-nearest neighbor regression (KNN), Random Forests (RF) or Support Vector Machines (SVM). These can be specified using the <em>method</em> argument.</p>
<p>For further details on the arguments that can be passed on the machine-learning methods, please see the documentation for <a href="https://cran.r-project.org/web/packages/randomForest/index.html">randomForest</a>, <a href="https://cran.r-project.org/web/packages/e1071/index.html">e0171</a> and <a href="https://cran.r-project.org/web/packages/FNN/index.html">FNN</a> packages.</p>
<p>By default, the number of training features that are used by each method are set at 50. This can adjusted by the user using the <em>num</em> argument to achieve a balance between desired imputation accuracy and computational costs.</p>
<p>Additionally, the “K”" or number of cross-validation iterations can be set using <em>folds</em>, which set at 10 by default.</p>
<p>Here, we perform a 10-fold cross-validation analysis using the entire GA training data sets using KNN method. We perform the analysis with both the processed and unprocessed GA protein-coding data for comparison</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Unprocessed training data</span>
CV.ga.gex1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/imirage.cv.loop.html">imirage.cv.loop</a></span>(<span class="dt">train_pcg =</span> ga.gex.<span class="dv">1</span>, <span class="dt">train_mir =</span> ga.mirna.<span class="dv">1</span>, <span class="dt">method =</span> <span class="st">"KNN"</span>, <span class="dt">target=</span><span class="st">"none"</span>)
<span class="co">#Processed training data</span>
CV.ga.gex2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/imirage.cv.loop.html">imirage.cv.loop</a></span>(<span class="dt">train_pcg =</span> ga.gex.<span class="dv">2</span>, <span class="dt">train_mir =</span> ga.mirna.<span class="dv">1</span>, <span class="dt">method =</span> <span class="st">"KNN"</span>, <span class="dt">target=</span><span class="st">"none"</span>)

<span class="co">#Unprocessed training data</span>
CV.hs.gex1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/imirage.cv.loop.html">imirage.cv.loop</a></span>(<span class="dt">train_pcg =</span> hiseq.gex.<span class="dv">1</span>, <span class="dt">train_mir =</span> hiseq.mirna.<span class="dv">1</span>, <span class="dt">method =</span> <span class="st">"KNN"</span>, <span class="dt">target=</span><span class="st">"none"</span>)
<span class="co">#Processed training data</span>
CV.hs.gex2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/imirage.cv.loop.html">imirage.cv.loop</a></span>(<span class="dt">train_pcg =</span> hiseq.gex.<span class="dv">2</span>, <span class="dt">train_mir =</span> hiseq.mirna.<span class="dv">1</span>, <span class="dt">method =</span> <span class="st">"KNN"</span>, <span class="dt">target=</span><span class="st">"none"</span>)


<span class="co">#Comparison of imputation accuracies between raw and processed data</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/boxplot">boxplot</a></span>(CV.ga.gex1[,<span class="dv">1</span>], CV.ga.gex2[,<span class="dv">1</span>], <span class="dt">names=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Raw"</span>, <span class="st">"Processed"</span>), <span class="dt">main=</span><span class="st">"GA CV accuracy"</span>)</code></pre></div>
<p><img src="imirage_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/boxplot">boxplot</a></span>(CV.hs.gex1[,<span class="dv">1</span>], CV.hs.gex2[,<span class="dv">1</span>], <span class="dt">names=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Raw"</span>, <span class="st">"Processed"</span>), <span class="dt">main=</span><span class="st">"Hiseq CV accuracy"</span>)</code></pre></div>
<p><img src="imirage_files/figure-html/unnamed-chunk-13-2.png" width="672"></p>
<p>Subsequently, the user can either select the miRNAs with good performance metrics in cross-validation analysis or perform the imputation using the full training data set with <code>imirage</code>. Generally, poor cross-validation metrics are a good indicator that the miRNA can be safely excluded from the prediction analysis in the independent test data set.</p>
<hr>
</div>
<div id="session-info" class="section level3">
<h3 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/sessionInfo">sessionInfo</a></span>()
<span class="co">#&gt; R version 3.6.1 (2019-07-05)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 10 x64 (build 18362)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=English_United States.1252 </span>
<span class="co">#&gt; [2] LC_CTYPE=English_United States.1252   </span>
<span class="co">#&gt; [3] LC_MONETARY=English_United States.1252</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                          </span>
<span class="co">#&gt; [5] LC_TIME=English_United States.1252    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] iMIRAGE_1.1         e1071_1.7-2         FNN_1.1.3          </span>
<span class="co">#&gt; [4] randomForest_4.6-14</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_1.0.1       rstudioapi_0.10  knitr_1.23       xml2_1.2.0      </span>
<span class="co">#&gt;  [5] magrittr_1.5     roxygen2_6.1.1   MASS_7.3-51.4    R6_2.4.0        </span>
<span class="co">#&gt;  [9] rlang_0.4.0      stringr_1.4.0    tools_3.6.1      xfun_0.8        </span>
<span class="co">#&gt; [13] class_7.3-15     htmltools_0.3.6  commonmark_1.7   yaml_2.2.0      </span>
<span class="co">#&gt; [17] digest_0.6.20    assertthat_0.2.1 rprojroot_1.3-2  pkgdown_1.3.0   </span>
<span class="co">#&gt; [21] crayon_1.3.4     fs_1.3.1         memoise_1.1.0    evaluate_0.14   </span>
<span class="co">#&gt; [25] rmarkdown_1.14   stringi_1.4.3    compiler_3.6.1   desc_1.2.0      </span>
<span class="co">#&gt; [29] backports_1.1.4</span></code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#introduction">Introduction</a><ul class="nav nav-pills nav-stacked">
<li><a href="#download-and-installation">Download and installation</a></li>
      <li><a href="#whats-included-in-this-package">Whats included in this package?</a></li>
      <li><a href="#report-issues">Report issues</a></li>
      </ul>
</li>
      <li>
<a href="#quick-start-guide">Quick Start Guide</a><ul class="nav nav-pills nav-stacked">
<li><a href="#step-1-match-protein-coding-genes-predictive-features-in-the-training-and-test-datasets-with-match-gex">Step 1: Match protein-coding genes (predictive features) in the training and test datasets with <code>match.gex</code></a></li>
      <li><a href="#step-2a-perform-cross-validation-analysis-with-imirage-cv-to-determine-the-expected-accuracy-of-imputing-a-mirna-of-interest-using-the-training-datasets">Step 2A: Perform cross-validation analysis with <code>imirage.cv</code> to determine the expected accuracy of imputing a miRNA of interest using the training datasets</a></li>
      <li><a href="#step-2b-perform-cross-validation-analysis-with-imirage-cv-loop-to-determine-the-expected-imputation-accuracy-of-entire-mirna-dataset-with">Step 2B: Perform cross-validation analysis with <code>imirage.cv.loop</code> to determine the expected imputation accuracy of entire miRNA dataset with</a></li>
      <li><a href="#step-3a-impute-expression-of-a-single-mirna-of-interest-using-imirage-in-the-test-dataset">Step 3A: Impute expression of a single miRNA of interest using <code>imirage</code> in the test dataset</a></li>
      <li><a href="#step-3b-use-imirage-to-impute-expression-of-all-mirnas-available-in-the-training-dataset">Step 3B: Use <code>imirage</code> to impute expression of all miRNAs available in the training dataset</a></li>
      </ul>
</li>
      <li>
<a href="#detailed-guide">Detailed guide</a><ul class="nav nav-pills nav-stacked">
<li><a href="#background">Background</a></li>
      <li><a href="#required-data">Required data</a></li>
      <li><a href="#workflow">Workflow</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Aritro Nath.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
